beacon-chain/db/kv/state.go

```diff
@@ -264,3 +298,73 @@ func createState(enc []byte) (*pb.BeaconState, error) {
 	}
 	return protoState, nil
 }
+
+// HighestSlotState returns the state with the highest slot from the db.
+func (k *Store) HighestSlotState(ctx context.Context) (*state.BeaconState, error) {
+	ctx, span := trace.StartSpan(ctx, "BeaconDB.HighestSlotState")
+	defer span.End()
+	var s *pb.BeaconState
+	err := k.db.View(func(tx *bolt.Tx) error {
+		slotBkt := tx.Bucket(slotsHasObjectBucket)
+		savedSlots := slotBkt.Get(savedStateSlotsKey)
+		highestIndex, err := bytesutil.HighestBitIndex(savedSlots)
+		if err != nil {
+			return err
+		}
+		highestSlot := highestIndex - 1
+		f := filters.NewFilter().SetStartSlot(uint64(highestSlot)).SetEndSlot(uint64(highestSlot))
+
+		keys, err := getBlockRootsByFilter(ctx, tx, f)
+		if err != nil {
+			return err
+		}
+		if len(keys) != 1 {
+			return errors.New("could not get one block root to get state")
+		}
+
+		stateBkt := tx.Bucket(stateBucket)
+		enc := stateBkt.Get(keys[0][:])
```

Comment:
What about forking conditions? How do you decide which block root is the canonical one?

Reply:
Forking won't result multiple block roots in a single slot unless a proposer commits a double propose. It could happen. I updated API to:
```
HighestSlotStates(ctx context.Context) ([]*state.BeaconState, error)
```